New documentation box design

OceanOak opened the pull request:
Changelog:

```
Internal improvements
-  Convert the documentation box to use Tailwind CSS.
-  Create a new documentation box design.
```
These are the first steps towards a new design for the documentation box.

Changes:
- Add function name and deprecated status to the documentation box.
- Display on error rail inside a tooltip.
- Create a new design for Type error + add Expected/Actual values to it.
- Use Tailwind CSS.

### Before -> After

**Function name + deprecated status**
![Normal fn](https://user-images.githubusercontent.com/87861924/210663852-1a9278b4-d2fc-4879-9f1a-98eb62809793.png)

**Deprecated:**
![Frame 21](https://user-images.githubusercontent.com/87861924/210666254-5eb61678-4876-433d-8b15-8eaa761dff89.png)

**On error rail:**
![onErrorRail](https://user-images.githubusercontent.com/87861924/210663240-f66cd7a1-f851-49ba-83da-8735c8ed7add.png)

**Type error:**
![TypeError](https://user-images.githubusercontent.com/87861924/210663473-3a466a58-7603-40db-8c30-780f7ed74fbf.png)



### These are just close-ups to help determine if any styling or spacing changes are needed.

**On error rail:** 
![Screenshot (11)](https://user-images.githubusercontent.com/87861924/210497684-ab43f0f7-aabf-458e-a5e9-48cbaa2512c3.png)

**Non-deprecated functions** 
![Screenshot (10)](https://user-images.githubusercontent.com/87861924/210497535-4c509001-e025-483a-9b87-a2b5d1c1f707.png)

**Deprecated functions + Type error**
![Screenshot (7)](https://user-images.githubusercontent.com/87861924/210497019-6b154410-ab44-45d1-8c01-3129dc1204ca.png)

pbiggar commented:
> Should the new Tooltip be moved from settingsComponent? (It is not part of settings so it doesn't make sense to keep it there). If so, where should we put it?

Yes, I'd love to have a Tooltip component that we could use anywhere (I tried to make one for the side bar but struggled with HTML layout issues).

I guess since it's a component, it should go in client/src/components.

OceanOak commented:
I've created a new tooltip that could be styled by the provider ( body, position, color, etc.)
1. Should we remove the one used in SettingsContributingView and use this one everywhere.
2. There is a file named Tooltips.res, it is used in the function space. Should we move the newly created tooltip there or should keep it in a seperate file?

pbiggar commented:
> **Questions:**
> 
> 1. I've created a new tooltip that could be styled by the provider ( body, position, color, etc.). Should we remove the one used in SettingsContributingView and use this one everywhere?

If it has some useful settings that are reused in Settings, then having it could be useful to reduce duplication. If not, then redoing that tooltip to use the component is valuable so we have less concepts/code to keep track of.


> 2. There is a file named Tooltips.res, it is used in the function space and the welcome to dark tutorial. Should we move the newly created tooltip there or should keep it in a seperate file?

Keep it separate I think. Tooltips.res is mostly about a tutorial, so we could rename it to Tutorial.res perhaps?

> 3. After converting PrettyDocs.res to use the Tailwind, I encountered some failing tests in TestPrettyDocs.res. The failing tests all involve a class name being expected, but a Tailwind class is returned instead. To resolve this issue, I directly used the Tailwind classes in the tests. However, I am uncertain if this is the most appropriate solution. If there is a better way to fix it, please let me know.
> 
> Example:
> 
> ```
> Expected : var  // a class name
> Actual : text-purple1 //tailwind
> ```

That feels like a good approach. It is possibly one of the downsides to tailwind, but it's probably not terrible.


> **Issues:**
> 
> 1. On error rail appears only when cursor is on complete function name, when it used to appear on AC and when the cursor is on the arguments as well.
>    Previously we've used `list{ViewErrorRailDoc.hintForFunction(f, sendToRail)}` to display it. Now because we moved it inside a tooltip I had to test if the function is on Rail or NoRail in order to display `(On Error Rail "i")`. Is there a better way to do it?

Do you feel it should appear on AC and when the cursor is on the arguments? If so, then we should keep it. I don't quite understand the code issue - presumably you can pass the errorrail status from the function to the tooltip?

> 2. `fluid_show_docs_for_command_on_selected_code` integration test is failing. Didn't manage to find a way to fix it any guidence here is appriciated.

Since there wasn't a video or a trace.zip stored in CI, the best approach is to:
- try it locally, but change the playwright config to always save a video or trace that you can replay.
- manually repeat the steps from the integration test to see if that finds the flaw 
- load the canvas locally after running the integration test to see its state (go to darklang.localhost:9000/a/test-WHATEVER-THE-INTEGRATION-TEST-NAME-IS

OceanOak commented:
> Do you feel it should appear on AC and when the cursor is on the arguments? If so, then we should keep it. I don't quite understand the code issue - presumably you can pass the errorrail status from the function to the tooltip?

I think it should when the cursor is on the arguments at least, because we rarely have the cursor on the function's name.

> Since there wasn't a video or a trace.zip stored in CI, the best approach is to:
> try it locally, but change the playwright config to always save a video or trace that you can replay.
> manually repeat the steps from the integration test to see if that finds the flaw
> load the canvas locally after running the integration test to see its state (go to darklang.localhost:9000/a/test-WHATEVER-THE-INTEGRATION-TEST-NAME-IS

This time, I didn't have to go through all of these steps,  but I still need to figure out how to run the integration tests because I can't run them inside the container or locally.

The fix was pretty straight forward I was just looking in the wrong place.
The issue is in `await page.waitForSelector(".documentation-box")` , which we no longer use. I commented it out to make sure that it was causing the issue. 

The question now is whether I should delete it or replace it with all of the Tailwind classes.

pbiggar commented:
> The question now is whether I should delete it or replace it with all of the Tailwind classes.

I think it's fine to leave in semantic classes for things like this (it also makes the html more readable when we are looking at it in the browser dev tools). Just so long as we don't have CSS attached to it.

OceanOak commented:
> I think it's fine to leave in semantic classes for things like this (it also makes the html more readable when we are looking at it in the browser dev tools). Just so long as we don't have CSS attached to it.

I couldn't use the Tailwind classes (I got issues with escaping some characters), so I've added an ID to the documentation box instead.

OceanOak commented:
> Do you feel it should appear on AC and when the cursor is on the arguments? If so, then we should keep it. I don't quite understand the code issue - presumably you can pass the errorrail status from the function to the tooltip?

To explain this better:
1. If we don't know if the function is on the rail, we display a generic message (this is what we display when the cursor is on the argument).
2. If we know if the function goes on the rail, we display a specific message.

I managed to make it appear when the cursor is on the argument. The problem is when the function isn't a function that goes to the error rail I am still getting an empty tooltip. Here is an example: 
![dark-Dark-Google-Chrome-2023-01](https://user-images.githubusercontent.com/87861924/211081162-a2b7657f-e0cc-46f0-97e6-e7ead0189976.gif)

pbiggar commented:
> I managed to make it appear when the cursor is on the argument. The problem is when the function isn't a function that goes to the error rail I am still getting an empty tooltip only when the cursor is on the argument. 

It shouldn't show the `(on ErrorRail (i))` there at all, since it's not on the rail.

OceanOak commented:
> It shouldn't show the `(on ErrorRail (i))` there at all, since it's not on the rail.
Yes, that's what I trying to achieve. and it is not showing when the cursor is on the function's name only when it is on the argument.

OceanOak commented:
Changes:

**Main.res** and **Tutorial.res** and **Entry.res** and **ViewUserFunction**
- Changed Tooltips to Tutorial because we renamed the Tooltips.res file to Tutorial.res 

**View.Res**
-converted the documentation box to Tailwind.
-added an ID for the documentation box to make the it easier to reference in the integration test
- changed Tooltips to Tutorial (because of the renaming)

**ViewErrorRailDoc.res**
-converted the error rail text that appears inside the tooltip to Tailwind and extracted the shared styles to avoid repeating the same code.

**SettingsCanvasesView.res** and **SettingsInviteView.res** and **SettingsView.res**
No changes in this file just formatting.

**SettingsViewComponents**
Fixed a comments and formatting.

**Tooltip.res**
A new file where I attempted to create a generic tooltip that we can use anywhere and can be styled by the provider (position, color,etc)

**FluidAutocomplete.res**
1. switched it to tailwind + changed the font 
2. type error
      - added expected/actual 
      - added a div that includes expected/actual and the error message.
      - extracted shared styles
      
![image](https://user-images.githubusercontent.com/87861924/211410149-d64f4880-d6fa-4a5e-8771-de827a3b05e8.png)


3. added a documentation header deprecationHeader that specifies whether the function is deprecated or not and the functions name
![image](https://user-images.githubusercontent.com/87861924/211410613-9017125e-30a3-489a-a38c-025c9dcc0eee.png)
![image](https://user-images.githubusercontent.com/87861924/211410664-a0b0e997-1320-40cf-aea2-06ff46c6ee79.png)

If the function is deprecated we've added a tooltip with a generic text about function's being deprecated (it is kind of ugly but I don't have an idea how to make it better yet).
![image](https://user-images.githubusercontent.com/87861924/211411181-c76878be-df1b-4aec-9cd6-c54af87bd002.png)

5. on error rail
moved on errorRail inside a tooltip
![image](https://user-images.githubusercontent.com/87861924/211413005-56ac21f1-ee5d-4706-b968-903568c4bd2f.png)

6. deprecationFooter
changed the styling
![image](https://user-images.githubusercontent.com/87861924/211412007-aa347e3d-2169-483b-8d38-9b4a260b3eaf.png)

7. Function's return type
-added an icon
![image](https://user-images.githubusercontent.com/87861924/211411396-71ea3065-3dc2-4054-9879-bc5fcec9d939.png)

**PrettyDocs.res**
-switched to tailwind 

**TestPrettyDocs.res**
-used tailwind instead of css class

**Tailwind.config.js**
-expanded the color palette + put the fontsizes in the right order

pbiggar commented on client/src/fluid/FluidAutocomplete.res:
```diff
@@ -763,55 +812,131 @@ let documentationForFunction = (
+    | ReplacedBy(name) => list{
+        Html.span(
+          list{tw(sharedStyle)},
+          list{Html.text("replaced by " ++ FQFnName.StdlibFnName.toString(name))},
```
Can you style the function name here?

pbiggar commented on client/src/fluid/FluidAutocomplete.res:
```diff
@@ -763,55 +812,131 @@ let documentationForFunction = (
+    | RenamedTo(name) => list{
+        Html.span(
+          list{tw(sharedStyle)},
+          list{Html.text("renamed to " ++ FQFnName.StdlibFnName.toString(name))},
```
Can you style the function name here?

pbiggar commented on client/src/fluid/FluidAutocomplete.res:
```diff
@@ -707,7 +710,11 @@ let selectUp = (a: t): t =>
+let fontSharedStyle = %twc("font-text tracking-wider")
+
 let typeErrorDoc = ({item, validity}: data): Vdom.t<AppTypes.msg> => {
+  let expected = Html.span(list{tw2(fontSharedStyle, %twc("mx-1"))}, list{Html.text("Expected: ")})
```
Having the green/orange text have the same foreground as background is a little jarring - can it be changed?

Also, it strikes me that this green/orange isn't from our main color palate - should it be changed to colors from our palate?

pbiggar commented on client/src/ui/PrettyDocs.res:
```diff
@@ -12,12 +12,19 @@ let codeEx = "^(.*)\\{\\{(.+)\\}\\}(.*)$"
 
 let nestedCodeBlock = Regex.regex(`{{[^}]+{{`)
 
+let \"type" = %twc("text-green")
```
You can calling this variable `typ` instead.

pbiggar commented on tailwind.config.js:
```diff
@@ -27,11 +27,14 @@ module.exports = {
 
       red: "#ab4642",
       orange: "#dc9656",
+      orange1: "#a67254",
```
Can you add a comment for orange1 like you did for purple1?

pbiggar reviewed (approved):
This all looks great. I have some comments but please address them as a followup PR so that this can be merged ASAP.
